cmd/migrate/main.go

```diff
@@ -339,27 +339,38 @@ func (m *chunkMover) moveChunks(ctx context.Context, threadID int, syncRangeCh <
 						keys = append(keys, key)
 						chks = append(chks, chk)
 					}
-					for retry := 10; retry >= 0; retry-- {
-						chks, err = f.FetchChunks(m.ctx, chks, keys)
-						if err != nil {
-							if retry == 0 {
-								log.Println(threadID, "Final error retrieving chunks, giving up:", err)
-								errCh <- err
-								return
+					finalChks := make([]chunk.Chunk, 0, len(chunks))
+					for i := range chks {
+						onechunk := []chunk.Chunk{chunks[i]}
+						onekey := []string{keys[i]}
+						var retry int
+						for retry = 10; retry >= 0; retry-- {
+							onechunk, err = f.FetchChunks(m.ctx, onechunk, onekey)
```

Comment:
Q: I have a feeling copying once chunk at a time may be super slow for the migration process? (additionally, now even single chunk failure would retry upto 10 times)

wondering does `f.FetchChunks` returns partial chunks and an non-nil error in that case you are describing?. 
[Looks like it though.](https://github.com/kavirajk/loki/blob/kavirajk%2Fx-rule-uid-grafana-managed-alerts/pkg/storage/chunk/client/util/parallel_chunk_fetch.go#L75-L76).

in that case, can we still use `batch` and use the partial chunks returned from the store? instead of ignoring those.

Reply:
Regarding speed, there's a disclaimer somewhere in here saying the migrate tool is not optimized, so yes, definitely a known issue, and this isn't helping.

However, regarding the one by one, I did digging into this, and this line winds up throwing out the partial fetch:
https://github.com/grafana/loki/blob/main/pkg/storage/chunk/fetcher/fetcher.go#L225

So, while I'd like to do what you are proposing, it requires more rework within Loki itself.

Reply:
Gotcha! thanks for the more info. Sounds good to me :+1: